core/src/main/java/com/linecorp/armeria/internal/metric/CaffeineMetricSupport.java

```diff
@@ -47,63 +59,149 @@ public static void setup(MeterRegistry registry, MeterId id, Cache<?, ?> cache)
+         */
+        private final double[] statsForGarbageCollected = new double[Type.count];
 
-            updateCacheStats(true);
+        CaffeineMetrics(MeterRegistry parent, MeterId id) {
+            this.parent = requireNonNull(parent, "parent");
+            this.id = requireNonNull(id, "id");
 
             final String requests = id.name("requests");
             parent.more().counter(requests, id.tags("result", "hit"), this,
-                                  new CacheStatFunction(() -> cacheStats.hitCount()));
+                                  func(HIT_COUNT, ref -> ref.cacheStats.hitCount()));
             parent.more().counter(requests, id.tags("result", "miss"), this,
-                                  new CacheStatFunction(() -> cacheStats.missCount()));
+                                  func(MISS_COUNT, ref -> ref.cacheStats.missCount()));
+            parent.more().counter(id.name("evictions"), id.tags(), this,
+                                  func(EVICTION_COUNT, ref -> ref.cacheStats.evictionCount()));
+            parent.more().counter(id.name("evictionWeight"), id.tags(), this,
+                                  func(EVICTION_WEIGHT, ref -> ref.cacheStats.evictionWeight()));
+            parent.gauge(id.name("estimatedSize"), id.tags(), this,
+                         func(null, ref -> ref.estimatedSize));
+        }
+
+        void add(Cache<?, ?> cache, Ticker ticker) {
+            synchronized (cacheRefs) {
+                for (CacheReference ref : cacheRefs) {
+                    if (ref.get() == cache) {
+                        // Do not aggregated more than once for the same instance.
+                        return;
+                    }
+                }
+
+                cacheRefs.add(new CacheReference(cache, ticker));
+            }
 
-            if (cache instanceof LoadingCache) {
+            if (cache instanceof LoadingCache && hasLoadingCache.compareAndSet(false, true)) {
+                // Add the following meters only for LoadingCache and only once.
                 final String loads = id.name("loads");
+
                 parent.more().counter(loads, id.tags("result", "success"), this,
-                                      new CacheStatFunction(() -> cacheStats.loadSuccessCount()));
+                                      func(LOAD_SUCCESS_COUNT, ref -> ref.cacheStats.loadSuccessCount()));
+
                 parent.more().counter(loads, id.tags("result", "failure"), this,
-                                      new CacheStatFunction(() -> cacheStats.loadFailureCount()));
+                                      func(LOAD_FAILURE_COUNT, ref -> ref.cacheStats.loadFailureCount()));
 
-                final DoubleSupplier totalLoadTimeSupplier;
-                totalLoadTimeSupplier = () -> cacheStats.totalLoadTime();
+                final ToDoubleFunction<CacheReference> totalLoadTimeSupplier;
+                totalLoadTimeSupplier = ref -> ref.cacheStats.totalLoadTime();
```

Comment:
nit: how about removing totalLoadTimeSupplier? or is this for the readability?

Reply:
Inlined totalLoadTimeSupplier. It was extracted before to avoid the column limit (Checkstyle)